Refactors IA-Powered Carousels (Supplants #2161: ia.IAEditionSearch)

mekarpeles opened the pull request:
<!-- What issue does this PR close? -->
Closes #2161 (incomplete stale refactor of IAEditionSearch)

<!-- What does this PR achieve? [feature|hotfix|fix|refactor] -->

This PR `refactor`s how Open Library queries Archive.org Elastic Search for books matching certain queries (which are then converted into a fetch of the desired Open Library editions w/ correct Archive.org availability information) 

### Testing
<!-- Steps for reviewer to reproduce/verify what this PR does/fixes. -->

- [x] rendering homepage carousels should work (all the carousels should be there, e.g. thrillers -- otherwise, this reflects an error in the Elastic Search queries)
- [x] clicking on the title/link of IA-powered homepage carousels on dev + production should produce an Archive.org search results page with comparable # of results 
- [x] tests for IAEditionSearch should pass
- [x] subject pages should render with correct availability information
- [x] homepage carousel infinite loading should work
- [x] IA-powered "similar books" + "more by author" should work on books page(s): e.g. https://dev.openlibrary.org/books/OL8870200M/A_Tale_of_Two_Cities

### Stakeholders
<!-- @ tag stakeholders of this bug -->
@cdrini @cclauss

mekarpeles committed:
Followup to 2161: ia.IAEditionSearch refactor

mekarpeles committed:
fixing linting

cclauss commented:
It would be good if this PR could ensure that [`openlibrary/plugins/upstream/tests/test_related_carousels.py`](https://travis-ci.org/github/internetarchive/openlibrary/jobs/662306587#L893) passes on both Py2 and Py3 as it does in #3177.

mekarpeles commented on openlibrary/tests/core/test_processors.py:
```diff
@@ -28,7 +28,7 @@ def _request(self, path, method=None, data=None):
             return web.storage(key=self.olids.get(olid))
 
-    def _get_backreferences(self):
+    def _get_backreferences(self, *args, **kwargs):
```
this technically fixes a different booklending_utils bug which makes tests fail on dev (because the booklending_utils lib is actually present)

mekarpeles commented:
> https://github.com/internetarchive/openlibrary/pull/3180#issuecomment-599019771

@cclauss I can barely tell what's going on in travis for python 3.8 linting on this branch

cclauss commented:
On your local machine run `make lint-diff`

cclauss commented:
At the very bottom of the Travis output:
openlibrary/core/ia.py:183:9: F401 'openlibrary.core.lending' imported but unused
openlibrary/plugins/worksearch/tests/test_worksearch.py:87:45: W605 invalid escape sequence '\:'
openlibrary/plugins/worksearch/tests/test_worksearch.py:93:46: W605 invalid escape sequence '\:'
openlibrary/tests/core/test_ia.py:2:1: F401 'infogami.config' imported but unused

cclauss commented on openlibrary/core/ia.py:
```diff
@@ -2,24 +2,239 @@
+        """
+        try:
+            import urllib2  # to port to requests
+            request = urllib2.Request(url)
```
```suggestion
            request = six.moves.urllib.request.Request(url)
```
And lose the import line above.

cclauss commented on openlibrary/core/ia.py:
```diff
@@ -2,24 +2,239 @@
+            # we can attribute requests to end-users
+            client_ip = web.ctx.env.get('HTTP_X_FORWARDED_FOR', 'ol-internal')
+            request.add_header('x-client-id', client_ip)
+            response = urllib2.urlopen(
```
```suggestion
            response =  six.moves.urllib.request.urlopen(
```

mekarpeles committed:
committing missing carousels page

mekarpeles committed:
Update openlibrary/core/ia.py

Co-Authored-By: Christian Clauss <cclauss@me.com>

mekarpeles committed:
Update openlibrary/core/ia.py

Co-Authored-By: Christian Clauss <cclauss@me.com>

cclauss commented:
Requires the same two imports as https://github.com/internetarchive/openlibrary/pull/3189#discussion_r393464813

mekarpeles committed:
fixing linting

cclauss commented:
Please rebase.  https://travis-ci.org/github/internetarchive/openlibrary/jobs/665700598#L584 was fixed in #3241

cdrini committed:
Code review comments

cdrini committed:
Refactor Edition to include set_availability method

This allows IAEditionSearch to set the availability, preventing unnecessary metadata calls.

cdrini committed:
Fix tests missing new loans__status__num_waitlist, etc.

cdrini committed:
Add a TODO docstring

cclauss commented on openlibrary/core/ia.py:
```diff
@@ -2,24 +2,252 @@
+        :return: the `response` content of the advancesearch
+        """
+        try:
+            # TODO switch to requests
```
Why postpone the shift to requests?

cclauss commented on openlibrary/core/ia.py:
```diff
@@ -2,24 +2,252 @@
+
+    @staticmethod
+    def _index_items_by_ocaid(items):
+        return dict((item['identifier'], item) for item in items)
```
Why pay the function call overhead?
```suggestion
        return {item['identifier']: item for item in items}
```

mekarpeles committed:
replacing legacy calls with new ia.IAEditionSearch

cclauss commented on openlibrary/core/models.py:
```diff
@@ -211,12 +211,12 @@ def canonicalize(edition):
+            ]
             if ol_covers:
-                return next(ol_covers)
+                return ol_covers[0]
```
The current implementation only processes the loop until the first element is created.  The proposed change creates the entire list and then only uses the first element.  Is this extra processing desired and useful?

tfmorris commented on openlibrary/core/ia.py:
```diff
@@ -2,24 +2,258 @@
+        'loans__status__last_loan_date'
+    ]
+
+    PRESET_QUERIES = {
```
A core library module seems like a strange place to store presets. I'd have thought they'd be in some calling code.

tfmorris commented on openlibrary/core/ia.py:
```diff
@@ -2,24 +2,258 @@
+        :rtype: dict
+        """
+        _limit = limit or cls.MAX_EDITIONS_LIMIT
+        return {
```
This seems like it belongs in a Solr helper library.

tfmorris commented on openlibrary/core/lending.py:
```diff
@@ -93,35 +91,17 @@ def setup(config):
     """
-    from openlibrary.plugins.openlibrary.home import CAROUSELS_PRESETS
-    query = CAROUSELS_PRESETS[query] if query in CAROUSELS_PRESETS else query
+    query = ia.IAEditionSearch.PRESET_QUERIES.get(query, query)
```
The previous module seems more logical to me.

tfmorris reviewed (changes requested):
It's pretty hard to review this without a high-level description of what the intended refactoring is meant to do. It looks like there were changes to the caching, but I'm not sure I followed them all.

Given that, my review focuses mostly on micro level syntactical stuff as opposed to whether the refactoring was done correctly at a macro level.
